daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/DataValue.scala

```diff
@@ -0,0 +1,209 @@
+  @inline implicit def toDataValue(v: JDouble): DataValueDouble = new DataValue(v)
+  @inline implicit def toDataValue(v: JBigInt): DataValueBigInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JString): DataValueString = new DataValue(v)
+  @inline implicit def toDataValue(v: URI): DataValueURI = new DataValue(v)
+  @inline implicit def toDataValue(v: JFloat): DataValueFloat = new DataValue(v)
+  @inline implicit def toDataValue(v: JByte): DataValueByte = new DataValue(v)
+  @inline implicit def toDataValue(v: JInt): DataValueInt = new DataValue(v)
+  @inline implicit def toDataValue(v: JShort): DataValueShort = new DataValue(v)
+  @inline implicit def toDataValue(v: DINode): DataValueDINode = new DataValue(v)
+
+  @inline implicit def toDataValue(v: Long): DataValueLong = new DataValue(v: JLong)
+  @inline implicit def toDataValue(v: Double): DataValueDouble = new DataValue(v: JDouble)
+  @inline implicit def toDataValue(v: Boolean): DataValueBool = new DataValue(v: JBoolean)
+  @inline implicit def toDataValue(v: Float): DataValueFloat = new DataValue(v: JFloat)
+  @inline implicit def toDataValue(v: Byte): DataValueByte = new DataValue(v: JByte)
+  @inline implicit def toDataValue(v: Int): DataValueInt = new DataValue(v: JInt)
+  @inline implicit def toDataValue(v: Short): DataValueShort = new DataValue(v: JShort)
+
+  @inline def unsafeFromAnyRef(v: AnyRef) = new DataValue(v)
+  @inline def unsafeFromMaybeAnyRef(v: Maybe[AnyRef]) = {
+    if (v.isDefined) {
+      new DataValue(v.get)
+    } else {
+      NoValue
+    }
+  }
+
+  val NoValue: DataValueEmpty = new DataValue(null)
+  val UseNilForDefault: DataValueUseNilForDefault = new DataValue(new UseNilForDefaultObj)
+
+  final protected class UseNilForDefaultObj {
+    override def toString = "UseNilForDefault"
+  }
+
+  @inline def assertValueIsNotDataValue(v: AnyRef) {
+
+    /*
+   *
+   * In our CompileExpressions classes, we use type variables declared as T <: AnyRef
+   *
+   * Ideally, we would have declared T <: DataValuePrimitive
+   * However, we need to be able to refer to something of the form Maybe[T].
+   * In theory, it should be possible to take a T <: DataValuePrimitive,
+   * and construct a T' <: DataValuePrimitiveNullable, such that T <: T'
+   * In practice, it does not appear to be easy to tell Scala's type system what we want to do,
+   * so we instead punt on this issue and require the caller to translate to/from AnyRef at the boundery.
+   *
+   * In theory, if a caller forgets to do so, and instead passes in a DataValue instead of AnyRef,
+   * the compiler would issue a type error because DataValue is an AnyVal type, and so does not inherit from AnyRef.
+   *
+   * In practice, Scala will "helpfully" box the DataValue for us, which causes all sorts of problems. For instance,
+   * x.asInstanceOf[String] does not work when x is a DataValueString (even though an unboxed DataValueString is literally just a String at runtime).
```

Comment:
Can we create an implicit conversion from DataValue to AnyRef that just fails? At runtime anyway we should be able to. Getting it to fail to compile would be great, but I think this full-service boxing is consider a feature.